docs/Using Randomness/v2-adhoc/design.md

```diff
@@ -0,0 +1,128 @@
+>
+> - If you are using v1, see the [VRF v1 guide](/docs/vrf/v1/introduction/).
+
+**Table of contents**
+
+- [Overview](#overview)
+- [Concepts](#concepts)
+  - [VRF Ad-hoc](#vrf-ad-hoc)
+  - [Request and Receive Data](#request-and-receive-data)
+    - [End To End Diagram](#end-to-end-diagram)
+    - [Explanation](#explanation)
+  - [Limits](#limits)
+
+## Overview
+
+This section explains how to generate random numbers using the ad-hoc method. This method doesn't require managing subscriptions and is more suited to one-off requests for randomness or situations where it is required to transfer the cost of VRF to end users as the cost is known at request time.
+
+## Concepts
+
+### VRF Ad-hoc
+
+Contrary to the [subscription method](/docs/vrf/v2/subscription/), users do not need to create subscriptions and pre-fund them: In fact, with the Ad-hoc method, users have to directly fund their consuming contracts with LINK tokens before requesting for randomness.
+
+For Chainlink VRF v2 to fulfill your requests, you must have a sufficient amount of LINK in your consuming contract. Gas cost calculation includes the following variables:
+
+- **Gas price:** The current gas price, which fluctuates depending on network conditions.
+
+- **Callback gas:** The amount of gas used for the callback request that returns your requested random values.
+
+- **Verification gas:** The amount of gas used to verify randomness on-chain.
+
+- **Wrapper overhead gas:** The amount of gas used by the VRF Wrapper contract (see [here](#request-and-receive-data) fore more details on the design and VRF v2 Wrapper contract).
+
+The gas price depends on current network conditions. The callback gas depends on your callback function, and the number of random values in your request. You define the limits you are willing to spend for the request with the following variable:
+
+- **Callback gas limit:** Specifies the maximum amount of gas you are willing to spend on the callback request. Define this limit by specifying the `callbackGasLimit` value in your request.
+
+> ℹ️ Note on transaction costs.
+>
+> Because the consuming contract directly pays for LINK, The cost is calculated during the request and not when the randomness is fulfilled (during the callback).
+> It is strongly recommended to test several times your callback function and correctly estimate the callback gas limit. In fact:
+>
+> - If the gas limit is underestimated: the callback will fail but your consuming contract is still charged for the work done to generate the requested random values.
+> - If the gas limit is overestimated: the callback function will be executed but you won't get refunded for the excess gas amount that you paid.
+>
+> Also, make sure that your consuming contracts are funded with enough LINK tokens to cover the transaction costs. In fact, if the consuming contract doesn't have enough LINK tokens then your request will revert.
```

Comment:
```suggestion
> Because the consuming contract directly pays the LINK for the request, the cost is calculated during the request and not during the callback when the randomness is fulfilled. Test your callback function to learn how to correctly estimate the callback gas limit.
>
> - If the gas limit is underestimated, the callback fails and the consuming contract is still charged for the work done to generate the requested random values.
> - If the gas limit is overestimated, the callback function will be executed but your contract is not refunded for the excess gas amount that you paid.
>
> Make sure that your consuming contracts are funded with enough LINK tokens to cover the transaction costs. If the consuming contract doesn't have enough LINK tokens, your request will revert.
```